src/lib/proxy-server.ts

```diff
@@ -0,0 +1,140 @@
+import { SiteServer } from 'src/site-server';
+import { loadUserData } from 'src/storage/user-data';
+
+let proxyServer: http.Server | null = null;
+let isProxyRunning = false;
+
+/**
+ * Gets the site details for a given domain by looking it up in user data and SiteServer
+ */
+async function getSiteByHost( domain: string ): Promise< SiteDetails | null > {
+	try {
+		const userData = await loadUserData();
+		const site = userData.sites.find(
+			( site ) => domainToASCII( site.customDomain ?? '' ) === domainToASCII( domain )
+		);
+		if ( site ) {
+			const server = SiteServer.get( site.id );
+			return server ? server.details : null;
+		}
+
+		return null;
+	} catch ( error ) {
+		console.error( 'Error looking up domain in user data:', error );
+		return null;
+	}
+}
+
+/**
+ * Attempts to start the proxy server on port 80
+ * This requires admin/root privileges
+ */
+export async function startProxyServer(): Promise< boolean > {
+	if ( isProxyRunning ) return true;
+
+	try {
+		return await startDomainProxy();
+	} catch ( error ) {
+		console.error( 'Failed to start proxy server:', error );
+		return false;
+	}
+}
+
+/**
+ * Starts the proxy server for sites with custom domains
+ */
+async function startDomainProxy(): Promise< boolean > {
+	try {
+		// Create proxy with additional options to preserve host header
+		const proxy = httpProxy.createProxyServer();
+
+		proxy.on( 'error', ( err, req, res ) => {
+			if ( res && res instanceof http.ServerResponse ) {
+				res.writeHead( 500 );
+				res.end( 'Proxy error: ' + err.message );
+			}
+		} );
+
+		proxyServer = http.createServer( async ( req, res ) => {
+			const host = req.headers.host?.split( ':' )[ 0 ]; // Remove port if present
+
+			// Look up the port directly from user data
+			if ( ! host ) {
+				console.log( 'No host header found' );
+				res.writeHead( 404 );
+				res.end( 'No host header found' );
+				return;
+			}
+
+			const site = await getSiteByHost( host );
+			if ( ! site ) {
+				console.log( `Domain not found: ${ host }` );
+				res.writeHead( 404 );
+				res.end( `Domain not found: ${ host }` );
+				return;
+			}
+
+			if ( ! site.running ) {
+				res.writeHead( 404 );
+				res.end( `The Studio site is currently stopped: ${ site.name }` );
+				return;
+			}
+
+			// Forward the request with the original host preserved
+			proxy.web( req, res, {
+				target: `http://localhost:${ site.port }`,
+				xfwd: true, // Pass along x-forwarded headers
+			} );
+		} );
+
+		await new Promise< void >( ( resolve, reject ) => {
+			proxyServer!
+				.listen( 80, () => {
+					console.log( `Proxy server started on port 80` );
+					isProxyRunning = true;
+					resolve();
+				} )
+				.on( 'error', ( err ) => {
+					console.error( `Error starting proxy server on port 80:`, err );
+					reject( err );
+				} );
+		} );
+
+		return true;
+	} catch ( error ) {
+		Sentry.captureException( error );
+		console.error( `Failed to start proxy server directly:`, error );
+		return false;
```

Comment:
```suggestion
		if ( error instanceof Error && error.message.includes( 'EADDRINUSE' ) ) {
			dialog.showMessageBox( { // imported from the `electron` module
				type: 'error',
				message: __( 'Failed to start custom domain proxy server' ),
				detail: __(
					'Another server is already running on port 80. For custom domains to work, please stop that server and then restart Studio.'
				),
				buttons: [ __( 'OK' ) ],
			} );

			return false;
		}

		Sentry.captureException( error );
		console.error( `Failed to start proxy server directly:`, error );
		return false;
```
While testing, I ran into a case where port 80 was already in use, which led to a pretty weird user experience (the site thumbnail in Studio was a 404 page from my other app). We should look for `EADDRINUSE` errors here and display an error modal.

My suggestion is far from perfect, but it's a start. Feel free to add or improve.

Reply:
Maybe we need a way to display error notifications in the UI in a more unified way. I guess that would do it for now.

Reply:
We typically use `showErrorMessageBox` from the IPC API for that, displaying an error modal like this one (modals for errors, notifications for successes). I agree it doesn't look very pretty in this case, but acknowledging the error is still better UX, IMO. I created https://github.com/Automattic/dotcom-forge/issues/10583 for iterating on the error message.